### PR TITLE
Deprecate `LightningLoggerBase.close`

### LINKED ISSUE #9421: Deprecate `LightningLoggerBase.close`
### Motivation

This is a follow up to https://github.com/PyTorchLightning/pytorch-lightning/discussions/9004#discussioncomment-1212966
and
https://github.com/PyTorchLightning/pytorch-lightning/issues/9037

The base logger API has `close` defined
https://github.com/PyTorchLightning/pytorch-lightning/blob/089ae9b3e82ddc31942e315294e31e48c0a899db/pytorch_lightning/loggers/base.py#L312-L314

This is only implemented on https://github.com/PyTorchLightning/pytorch-lightning/blob/089ae9b3e82ddc31942e315294e31e48c0a899db/pytorch_lightning/loggers/test_tube.py#L197-L204

Given the test tube logger has since been deprecated, we can also deprecate this method off the base API as it's very unclear what the difference is between save/close/finalize currently. 

This function is never called by the Trainer either, so deprecating it from the base logger API has minimal changes for users.

### Pitch

- Deprecate `close` off the base API in v1.5
- Remove it from the API in v1.7

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #9421
deprecate `LightningLoggerBase.close`

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/base.py [hunk 1]
```diff
@@ -391,10 +387,6 @@ def finalize(self, status: str) -> None:
         for logger in self._logger_iterable:
             logger.finalize(status)
 
-    def close(self) -> None:
-        for logger in self._logger_iterable:
-            logger.close()
-
```
Comment (ananthsub) [inline pytorch_lightning/loggers/base.py:397]: these can't be deleted outright. we have to first mark these as deprecated with a clear message for users, and allow time for users to migrate. https://pytorch-lightning.readthedocs.io/en/latest/governance.html#api-evolution

**Comment by edward-io** [comment]
Looks great, you can convert this into a non-draft PR when you're ready.

File: CHANGELOG.md [hunk 2]
```diff
@@ -197,6 +197,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated passing `process_position` to the `Trainer` constructor in favor of adding the `ProgressBar` callback with `process_position` directly to the list of callbacks ([#9222](https://github.com/PyTorchLightning/pytorch-lightning/pull/9222))
 
 
+- Deprecated `LightningLoggerBase.close`, `LoggerCollection.close` in favor of `LightningLoggerBase.save`, `LoggerCollection.save` ([#9422](https://github.com/PyTorchLightning/pytorch-lightning/pull/9422))
```
Comment (ananthsub) [inline CHANGELOG.md:200]: why's this duplicated? 

also @awaelchli @edward-io is there consensus on whether we should recommend the usage of `save` vs. `finalize` here? 
```suggestion
- Deprecated `LightningLoggerBase.close` in favor of `LightningLoggerBase.save` ([#9422](https://github.com/PyTorchLightning/pytorch-lightning/pull/9422))
```

Comment (jjenniferdai) [reply CHANGELOG.md:200]: - "duplicated": `LightningLoggerBase` and `LoggerCollection` ?
- yea I was wondering this too. I picked `save` here since `LightningLoggerBase.close` just calls `self.save` for identical functionality (v.s. `finalize` could've been re-implemented differently)

**Comment by jjenniferdai** [comment]
trying to figure out the make-docs failure. ([found](https://github.com/PyTorchLightning/pytorch-lightning/actions/workflows/docs-checks.yml) similar error in #9366 but the suggestion doesn't seem to apply here)

File: CHANGELOG.md (hunk 2 above)
Comment (ananthsub) [reply CHANGELOG.md:200]: > "duplicated": LightningLoggerBase and LoggerCollection ?

ah sorry, didn't read carefully!

File: pytorch_lightning/loggers/base.py [hunk 3]
```diff
@@ -310,7 +311,15 @@ def finalize(self, status: str) -> None:
         self.save()
 
     def close(self) -> None:
-        """Do any cleanup that is necessary to close an experiment."""
+        """Do any cleanup that is necessary to close an experiment.
+
+        .. deprecated:: v1.5     This method was deprecated in v1.5 and will be removed in v1.7     in favor of
+        `LightningLoggerBase.save`
```
Comment (ananthsub) [inline pytorch_lightning/loggers/base.py:317]: for the docs error, maybe there's an issue with the spacing here?

Comment (jjenniferdai) [reply pytorch_lightning/loggers/base.py]: this maybe does seem to be the issue? c707a9f [passes](https://github.com/PyTorchLightning/pytorch-lightning/runs/3563829458) but 5c77e69 somehow [does not](https://github.com/PyTorchLightning/pytorch-lightning/runs/3563863729)

File: pytorch_lightning/loggers/base.py [hunk 4]
```diff
@@ -310,7 +311,16 @@ def finalize(self, status: str) -> None:
         self.save()
 
     def close(self) -> None:
-        """Do any cleanup that is necessary to close an experiment."""
+        """
+        .. deprecated:: v1.5
+          This method was deprecated in v1.5 and will be removed in v1.7
+          in favor of `LightningLoggerBase.save`
+
+        """
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:319]:
```suggestion
        """
        .. deprecated:: v1.5
            This method was deprecated in v1.5 and will be removed in v1.7
            in favor of `LightningLoggerBase.save`
        """
```
also, generally we would keep the documentation in place during the deprecation phase

File: CHANGELOG.md (hunk 2 above)
Comment (awaelchli) [reply CHANGELOG.md:200]: normally, saving and closing would be different concepts. So I would favor finalize as the replacement. Unfortunately I don't know the original intentions of the authors behind these methods.

Comment (edward-io) [reply CHANGELOG.md:200]: Agreed that finalize would be the closest thing to close.

File: tests/loggers/test_base.py [hunk 5]
```diff
@@ -48,9 +48,9 @@ def test_logger_collection():
     mock1.agg_and_log_metrics.assert_called_once_with({"test": 2.0}, 4)
     mock2.agg_and_log_metrics.assert_called_once_with({"test": 2.0}, 4)
 
-    logger.close()
-    mock1.close.assert_called_once()
-    mock2.close.assert_called_once()
+    logger.finalize()
```
Comment (ananthsub) [inline tests/loggers/test_base.py:51]: this is missing the status: https://github.com/PyTorchLightning/pytorch-lightning/blob/7ca038b83f9d326b03944e8ffb308f22d0ea241e/pytorch_lightning/loggers/base.py#L304-L310

from the trainer, this seems to always be `"success"`: https://github.com/PyTorchLightning/pytorch-lightning/blob/7ca038b83f9d326b03944e8ffb308f22d0ea241e/pytorch_lightning/trainer/trainer.py#L1270

having `finalize` in loggers vs `teardown` everywhere else is another quirk which we could address in a separate issue/PR

File: pytorch_lightning/loggers/base.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loggers/base.py]: yep, just make the indentation like in all the other places and it will be good :)

Comment (jjenniferdai) [reply pytorch_lightning/loggers/base.py]: haven't been able to figure why docformatter keeps [modifying](https://github.com/PyTorchLightning/pytorch-lightning/pull/9422/commits/6d12d69876e51f394807223512d15824f8aefcfb) the correct matching format that actually [passes make-docs](https://github.com/PyTorchLightning/pytorch-lightning/runs/3569055223) :(

Comment (awaelchli) [reply pytorch_lightning/loggers/base.py]: oh wow, that seems to be a pretty bad bug. not sure what we can do about it, will ask around

File: pytorch_lightning/loggers/base.py [hunk 6]
```diff
@@ -310,7 +311,15 @@ def finalize(self, status: str) -> None:
         self.save()
 
     def close(self) -> None:
-        """Do any cleanup that is necessary to close an experiment."""
+        """Do any cleanup that is necessary to close an experiment.
+
+        .. deprecated:: v1.5     This method is deprecated in v1.5 and will be removed in v1.7.     Please use
+        `LightningLoggerBase.finalize` instead.
+        """
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:318]:
```suggestion
        """
        .. deprecated:: v1.5
            This method is deprecated in v1.5 and will be removed in v1.7.
            Please use `LightningLoggerBase.finalize` instead.
        
        Do any cleanup that is necessary to close an experiment.
        """
```
trying to move it at the top, lets see

File: pytorch_lightning/loggers/base.py [hunk 7]
```diff
@@ -310,7 +311,15 @@ def finalize(self, status: str) -> None:
         self.save()
 
     def close(self) -> None:
-        """Do any cleanup that is necessary to close an experiment."""
+        """.. deprecated:: v1.5 This method is deprecated in v1.5 and will be removed in v1.7. Please use
+        `LightningLoggerBase.finalize` instead.
+
+        Do any cleanup that is necessary to close an experiment.
+        """
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:318]:
```suggestion
        """Do any cleanup that is necessary to close an experiment.
        
        .. deprecated:: v1.5
           This method is deprecated in v1.5 and will be removed in v1.7. 
           Please use `LightningLoggerBase.finalize` instead.
        """
```

File: pytorch_lightning/loggers/base.py [hunk 8]
```diff
@@ -310,7 +311,15 @@ def finalize(self, status: str) -> None:
         self.save()
 
     def close(self) -> None:
-        """Do any cleanup that is necessary to close an experiment."""
+        """Do any cleanup that is necessary to close an experiment.
+
+        .. deprecated:: v1.5    This method is deprecated in v1.5 and will be removed in v1.7.    Please use
+        `LightningLoggerBase.finalize` instead.
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py:317]:
```suggestion
        .. deprecated:: v1.5
            This method is deprecated in v1.5 and will be removed in v1.7. Please use ``LightningLoggerBase.finalize`` instead.
```

File: pytorch_lightning/loggers/base.py (hunk 3 above)
Comment (Borda) [reply pytorch_lightning/loggers/base.py]: not sure what is the question?

Comment (awaelchli) [reply pytorch_lightning/loggers/base.py]: the docstring autoformatter is breaking the format

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/loggers/base.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loggers/base.py]: Looks like we can't do anything about it right now, so let's merge it anyway. The message is there, only the formatting is a bit broken

Comment (carmocca) [reply pytorch_lightning/loggers/base.py]: It won't get formatted if there's one more line

```python
def close():
    """Do any cleanup that is necessary to close an experiment.

    Having this line avoids the formatting.

    .. deprecated:: v1.5
        This method is deprecated in v1.5 and will be removed in v1.7.
        Please use ``something something`` instead.
    """
```

Worth opening an issue in https://github.com/myint/docformatter though

Comment (jjenniferdai) [reply pytorch_lightning/loggers/base.py]: thanks everyone! hm I added an extra line which does indeed do the trick, if it looks ok (filed https://github.com/myint/docformatter/issues/78)
